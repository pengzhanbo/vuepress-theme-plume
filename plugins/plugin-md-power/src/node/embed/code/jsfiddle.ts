/**
 * @[jsfiddle](user/id)
 * @[jsfiddle theme="dark" tab="js,css,html,result"](user/id)
 */
import type { PluginWithOptions } from 'markdown-it'
import type { JSFiddleTokenMeta } from '../../../shared/index.js'
import { parseRect } from '../../utils/parseRect.js'
import { resolveAttrs } from '../../utils/resolveAttrs.js'
import { createEmbedRuleBlock } from '../createEmbedRuleBlock.js'

export const jsfiddlePlugin: PluginWithOptions<never> = (md) => {
  createEmbedRuleBlock<JSFiddleTokenMeta>(md, {
    type: 'jsfiddle',
    syntaxPattern: /^@\[jsfiddle([^\]]*)\]\(([^)]*)\)/,
    meta([, info = '', source]) {
      const { attrs } = resolveAttrs(info)
      const [user, id] = source.split('/')

      return {
        width: attrs.width ? parseRect(attrs.width) : '100%',
        height: attrs.height ? parseRect(attrs.height) : '400px',
        user,
        id,
        title: attrs.title || 'JS Fiddle',
        tab: attrs.tab?.replace(/\s+/g, '') || 'js,css,html,result',
        theme: attrs.theme || 'dark',
      }
    },
    content: ({ title = 'JS Fiddle', height, width, user, id, tab, theme }) => {
      theme = theme === 'dark' ? '/dark/' : ''

      const link = `https://jsfiddle.net/${user}/${id}/embedded/${tab}${theme}`
      const style = `width:${width};height:${height};margin:16px auto;border:none;border-radius:5px;`
      return `<iframe class="js-fiddle-iframe-wrapper" style="${style}" title="${title}" src="${link}" allowfullscreen="true" allowpaymentrequest="true"></iframe>`
    },
  })
}
