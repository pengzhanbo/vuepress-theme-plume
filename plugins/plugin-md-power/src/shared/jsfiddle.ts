import type { SizeOptions } from './size'

export interface JSFiddleTokenMeta extends SizeOptions {
  source: string
  title?: string
  theme?: string
  tab?: string
}
