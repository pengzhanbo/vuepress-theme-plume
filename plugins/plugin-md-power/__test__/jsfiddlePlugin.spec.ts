import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { jsfiddlePlugin } from '../src/node/embed/code/jsfiddle.js'

function createMarkdown() {
  return MarkdownIt().use((md) => {
    md.block.ruler.before('code', 'import_code', () => false)
    md.renderer.rules.import_code = () => ''
  }).use(jsfiddlePlugin)
}

describe('codeSandboxPlugin', () => {
  it('should work', () => {
    const md = createMarkdown()
    const code = `\
@[jsfiddle](user/id)

@[jsfiddle](user/id)

@[jsfiddle theme="light"](user/id)

@[jsfiddle title="xxx" width="100%" height="500px" theme="dark" tab="js,css,html,result"](user/id)
`

    expect(md.render(code)).toMatchSnapshot()
  })

  it('should not work', () => {
    const md = createMarkdown()
    const code = `\
@[jsfiddle]()

@[jsfiddle]xxx

@[jsfiddle](
`

    expect(md.render(code)).toMatchSnapshot()
  })
})
