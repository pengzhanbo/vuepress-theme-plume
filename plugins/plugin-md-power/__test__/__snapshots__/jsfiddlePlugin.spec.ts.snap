// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codeSandboxPlugin > should not work 1`] = `
"<iframe class="js-fiddle-iframe-wrapper" style="width:100%;height:400px;margin:16px auto;border:none;border-radius:5px;" title="JS Fiddle" src="https://jsfiddle.net//undefined/embedded/js,css,html,result/dark/" allowfullscreen="true" allowpaymentrequest="true"></iframe><p>@[jsfiddle]xxx</p>
<p>@[jsfiddle](</p>
"
`;

exports[`codeSandboxPlugin > should work 1`] = `"<iframe class="js-fiddle-iframe-wrapper" style="width:100%;height:400px;margin:16px auto;border:none;border-radius:5px;" title="JS Fiddle" src="https://jsfiddle.net/user/id/embedded/js,css,html,result/dark/" allowfullscreen="true" allowpaymentrequest="true"></iframe><iframe class="js-fiddle-iframe-wrapper" style="width:100%;height:400px;margin:16px auto;border:none;border-radius:5px;" title="JS Fiddle" src="https://jsfiddle.net/user/id/embedded/js,css,html,result/dark/" allowfullscreen="true" allowpaymentrequest="true"></iframe><iframe class="js-fiddle-iframe-wrapper" style="width:100%;height:400px;margin:16px auto;border:none;border-radius:5px;" title="JS Fiddle" src="https://jsfiddle.net/user/id/embedded/js,css,html,result" allowfullscreen="true" allowpaymentrequest="true"></iframe><iframe class="js-fiddle-iframe-wrapper" style="width:100%;height:500px;margin:16px auto;border:none;border-radius:5px;" title="xxx" src="https://jsfiddle.net/user/id/embedded/js,css,html,result/dark/" allowfullscreen="true" allowpaymentrequest="true"></iframe>"`;
